File: app/src/main/java/org/sco/movieratings/adapter/MoviePreviewAdapter.java
package org.sco.movieratings.adapter;

import java.util.ArrayList;
import java.util.List;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import org.sco.movieratings.R;
import org.sco.movieratings.data.models.Preview;

public class MoviePreviewAdapter extends RecyclerView.Adapter<MoviePreviewAdapter.ViewHolder> {
    private static final String LOG_TAG = MoviePreviewAdapter.class.getSimpleName();

    private ArrayList<Preview> mPreviews;
    private final Callback mCallback;

    public interface Callback {
        void view(Preview preview);
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        private ImageView playButton;
        private TextView trailer;

        public ViewHolder(View v) {
            super(v);
            playButton = (ImageView) v.findViewById(R.id.playButton);
            trailer = (TextView) v.findViewById(R.id.trailerName);
        }
    }

    public MoviePreviewAdapter(ArrayList<Preview> previews, Callback callback) {
        this.mPreviews = previews;
        mCallback = callback;
    }

    public void add(List<Preview> previews) {
        mPreviews.clear();
        mPreviews.addAll(previews);
        notifyDataSetChanged();
    }

    public ArrayList<Preview> getPreviews() {
        return mPreviews;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.preview_list_item, parent, false);
        ViewHolder vh = new ViewHolder(v);
        return vh;
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        final Preview preview = mPreviews.get(position);

        holder.trailer.setText(preview.getName());

        holder.playButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mCallback.view(preview);
            }
        });
    }

    @Override
    public int getItemCount() {
        return mPreviews.size();
    }

}

File: app/src/main/java/org/sco/movieratings/adapter/MoviePreviewAdapter.kt
package org.sco.movieratings.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import org.sco.movieratings.R
import org.sco.movieratings.api.models.Preview

private const val TAG = "MoviePreviewAdapter"

class MoviePreviewAdapter(val previews: ArrayList<Preview>, private val callback: Callback) : RecyclerView.Adapter<MoviePreviewAdapter.MoviePreviewViewHolder>() {
    interface Callback {
        fun view(preview: Preview)
    }
    class MoviePreviewViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var playButton: ImageView = view.findViewById(R.id.playButton)
        var trailer: TextView = view.findViewById(R.id.trailerName)
    }

    fun add(previewsList: List<Preview>) {
        previews.clear()
        previews.addAll(previewsList)
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return previews.size
    }

    override fun onBindViewHolder(holder: MoviePreviewViewHolder, position: Int) {
        val preview: Preview = previews[position]
        holder.trailer.text =
            holder.itemView.resources.getString(R.string.preview_title, preview.site, preview.name)
        holder.playButton.setOnClickListener {
            callback.view(preview)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MoviePreviewViewHolder {
        return MoviePreviewViewHolder(LayoutInflater.from(parent.context)
            .inflate(R.layout.preview_list_item, parent, false))
    }

}